### PR TITLE
Infected user functionality

### PR DESCRIPTION
**Pull request by palwashamalik**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -4,12 +4,11 @@ class ApplicationController < ActionController::Base
 
   def configure_permitted_parameters
     devise_parameter_sanitizer.permit(:sign_up) do |u|
-      u.permit(:name, :email, :password, :age, :gender, :infected, :avatar,
-               items_attributes: %i[id points item_name quantity _destroy],
-               locations_attributes: %i[latitude longitude])
+      u.permit(:name, :email, :password, :age, :gender, :infected, :avatar, locationhistory: [:longitude, :latitude, :created_at],
+              items_attributes: %i[id points itemname quantity _destroy])
```
Comment (MishalNaeem) [inline app/controllers/application_controller.rb]: change itenname to item_name

File: app/controllers/locations_controller.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 class LocationsController < ApplicationController
   def index
-    @locations = current_user.locations.order('created_at DESC')
+    @locationhistory = current_user.locationhistory
```
Comment (MishalNaeem) [inline app/controllers/locations_controller.rb]: location_history instead of locationhistory

File: app/controllers/registrations_controller.rb [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
-  def edit
-    @location = resource.locations.new
-    render :edit
-  end
 
   def update_resource(resource, params)
+    if @user.locationhistory
+      locations = []
+      locations.push(@user.locationhistory)
+      locations.push(params[:locationhistory])
```
Comment (MishalNaeem) [inline app/controllers/registrations_controller.rb]: why we need to push user history separately and then params history?

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
+class UsersController < ApplicationController
+  before_action :find_user
+
+  def upvote
+    @user = User.find(params[:id])
```
Comment (MishalNaeem) [inline app/controllers/users_controller.rb]: use find_user method instead of find user here

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    @user.upvote_from current_user if @user.get_upvotes.size < 5
```
Comment (MishalNaeem) [inline app/controllers/users_controller.rb:5]: Can you please elaborate the use of this code

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    end
+    redirect_to root_path
+  end
+
+  def show
+    @items = @user.items
+  end
+
+  def downvote
+    @user = User.find(params[:id])
```
Comment (MishalNaeem) [inline app/controllers/users_controller.rb]: find_method used

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -1,5 +1,7 @@
 module ApplicationHelper
-  def image_helper(image)
-    image&.attached? ? image : 'default.jpeg'
+  def location_array(arr)
+    return arr if arr.kind_of?(Array)
+
+    return [arr]
   end
```
Comment (MishalNaeem) [inline app/helpers/application_helper.rb]: use this code in separate file  named locationHelper and use it

File: app/models/item.rb [hunk 8]
```diff
@@ -1,6 +1,9 @@
 class Item < ApplicationRecord
   belongs_to :user, optional: true
 
+<<<<<<< HEAD
   validates :points, numericality: { only_integer: true, greater_than_or_equal_to: 0 }, presence: true
   validates :item_name, presence: true
+=======
+>>>>>>> 6bcb94a (added item model)
```
Comment (MishalNaeem) [inline app/models/item.rb]: rebasing issue

File: db/migrate/20220725071959_acts_as_votable_migration.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    create_table :votes do |t|
+
+      t.references :votable, :polymorphic => true
+      t.references :voter, :polymorphic => true
+
+      t.boolean :vote_flag
+      t.string :vote_scope
+      t.integer :vote_weight
+
+      t.timestamps
```
Comment (MishalNaeem) [inline db/migrate/20220725071959_acts_as_votable_migration.rb]: Add proper DB level validations

File: db/migrate/20220727120103_add_location_history_to_users.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class AddLocationHistoryToUsers < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :locationhistory, :jsonb
```
Comment (MishalNaeem) [inline db/migrate/20220727120103_add_location_history_to_users.rb]: location_history
add proper db level validations

File: db/migrate/20220728130203_drop_locations.rb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+class DropLocations < ActiveRecord::Migration[5.2]
+  def change
+    drop_table :locations
+
```
Comment (MishalNaeem) [inline db/migrate/20220728130203_drop_locations.rb]: remove this line

**Review by MishalNaeem — CHANGES_REQUESTED** [review]

File: app/controllers/registrations_controller.rb (hunk 3 above)
Comment (palwashamalik) [reply app/controllers/registrations_controller.rb]: to put previous user locations and new location to an array

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (palwashamalik) [reply app/controllers/users_controller.rb:5]: if the upvotes(which means flags for infected user) is less then five then it counts the upvote

File: db/migrate/20220725071959_acts_as_votable_migration.rb (hunk 9 above)
Comment (palwashamalik) [reply db/migrate/20220725071959_acts_as_votable_migration.rb]: this is the migration created by votable gem. should i add db validations to it?

Comment (MishalNaeem) [reply db/migrate/20220725071959_acts_as_votable_migration.rb]: yes add it here too

**Review by MishalNaeem — APPROVED** [review]

**Pull request merged**